File: LeetCode/面试经典 150 题/双指针/392. 判断子序列/392.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    bool isSubsequence(string s, string t) {
        int n = t.size();
        int p = 0;
        for (auto it = s.begin(); it != s.end(); it++) {
            const char &c = *it;
            while (p < n && t[p] != c) { p++; }
            if (p >= n) { return false; }
            else { p++; }
        }
        return true;
    }
};

int main() {
    Solution s;
//    auto ans = s.isSubsequence("abc", "ahbgdc");
    auto ans = s.isSubsequence("axc", "ahbgdc");
    printf("%d", ans);
    return 0;
}

File: LeetCode/LeetCode 75/双指针/392. 判断子序列/392.java
class Solution {
    public boolean isSubsequence(String s, String t) {
        int tp = 0;
        for (int sp = 0; sp < s.length(); sp++) {
            while (tp < t.length() && s.charAt(sp) != t.charAt(tp)) { tp++; }
            if (tp >= t.length()) {
                return false;
            }
            tp++;
        }
        return true;
    }
}